### PR TITLE
skip broken symlinks when hashing

### LINKED ISSUE #288: Directory type fails to hash when annexed data is not local
Consider a tool to initialize a pybids database to reuse in future nodes:

```Python
@pydra.mark.task
def initialize_db(
        bids_dir: Directory,
        derivatives: ty.List[Directory],
        ignore: ty.List,
        force_index: ty.List,
        database_path: str,
        reset_database: bool) -> File:
    from bids import BIDSLayout
    BIDSLayout(bids_dir, derivatives=derivatives, ignore=ignore, force_index=force_index,
               database_path=database_path, reset_database=reset_database)
    return database_path
```

If you run this on a BIDS dataset with some missing contents, you will get something like:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-72bd956ec5a1> in <module>
----> 1 dbfile = initdb(bids_dir='/data/out/qnl/repeat_change/sourcedata', derivatives=['/data/out/qnl/repeat_change'], ignore=['.git', '.datalad', 'derivatives', 'sourcedata'], force_index=[], database_path='/tmp/qnlbids.db', reset_database=False)

~/anaconda3/envs/pydra-dev/lib/python3.8/site-packages/pydra/engine/core.py in __call__(self, submitter, plugin, rerun, **kwargs)
    365                     "TODO: linear workflow execution - assign submitter or plugin for now"
    366                 )
--> 367             res = self._run(rerun=rerun, **kwargs)
    368         return res
    369 

~/anaconda3/envs/pydra-dev/lib/python3.8/site-packages/pydra/engine/core.py in _run(self, rerun, **kwargs)
    371         self.inputs = attr.evolve(self.inputs, **kwargs)
    372         self.inputs.check_fields_input_spec()
--> 373         checksum = self.checksum
    374         lockfile = self.cache_dir / (checksum + ".lock")
    375         # Eagerly retrieve cached - see scenarios in __init__()

~/anaconda3/envs/pydra-dev/lib/python3.8/site-packages/pydra/engine/core.py in checksum(self)
    228             (before the tasks have inputs etc.)
    229         """
--> 230         input_hash = self.inputs.hash
    231         if self.state is None:
    232             self._checksum = create_checksum(self.__class__.__name__, input_hash)

~/anaconda3/envs/pydra-dev/lib/python3.8/site-packages/pydra/engine/specs.py in hash(self)
     51             if getattr(self, field.name) is attr.NOTHING:
     52                 continue
---> 53             inp_dict[field.name] = hash_value(
     54                 value=getattr(self, field.name), tp=field.type, metadata=field.metadata
     55             )

~/anaconda3/envs/pydra-dev/lib/python3.8/site-packages/pydra/engine/helpers.py in hash_value(value, tp, metadata)
    447         metadata = {}
    448     if isinstance(value, (tuple, list)):
--> 449         return [hash_value(el, tp, metadata) for el in value]
    450     elif isinstance(value, dict):
    451         dict_hash = {k: hash_value(v, tp, metadata) for (k, v) in value.items()}

~/anaconda3/envs/pydra-dev/lib/python3.8/site-packages/pydra/engine/helpers.py in <listcomp>(.0)
    447         metadata = {}
    448     if isinstance(value, (tuple, list)):
--> 449         return [hash_value(el, tp, metadata) for el in value]
    450     elif isinstance(value, dict):
    451         dict_hash = {k: hash_value(v, tp, metadata) for (k, v) in value.items()}

~/anaconda3/envs/pydra-dev/lib/python3.8/site-packages/pydra/engine/helpers.py in hash_value(value, tp, metadata)
    464             and "container_path" not in metadata
    465         ):
--> 466             return hash_dir(value)
    467         else:
    468             return value

~/anaconda3/envs/pydra-dev/lib/python3.8/site-packages/pydra/engine/helpers_file.py in hash_dir(dirpath, crypto, ignore_hidden_files, ignore_hidden_dirs, raise_notfound)
    140             if ignore_hidden_files and filename.startswith("."):
    141                 continue
--> 142             this_hash = hash_file(dpath / filename)
    143             file_hashes.append(this_hash)
    144 

~/anaconda3/envs/pydra-dev/lib/python3.8/site-packages/pydra/engine/helpers_file.py in hash_file(afile, chunk_len, crypto, raise_notfound)
     75     if not Path(afile).is_file():
     76         if raise_notfound:
---> 77             raise RuntimeError('File "%s" not found.' % afile)
     78         return None
     79 

RuntimeError: File "/data/out/qnl/repeat_change/.datalad/environments/fmriprep-20-1-0/image" not found.
```

Would it make sense to treat missing links as 0-length files for the purpose of hashing? Or detect git repositories and use git hashes instead?

### PR DESCRIPTION
**Pull request by nicolocin**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Summary

Fix #288 

## Checklist

- [x] All tests passing
- [x] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [x] My code follows the code style of this project
(we are using `black`: you can `pip install pre-commit`,
run `pre-commit install` in the `pydra` directory
and `black` will be run automatically with each commit)

## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

**Comment by djarecka** [comment]
@nicolocin - if file is just a link it can be skipped when you calculate the hash_value of directory. But if a file is an input field, the hash_file function should still raise an error

**Comment by nicolocin** [comment]
@djarecka ohh my bad, it's only used in `hash_value` and only existing files/dirs can be used in `hash_file` and `hash_dir`.  I'll simplify it

**Comment by djarecka** [comment]
@nicolocin - could we add test to check if this is what we want:
1. symlink is provided as an input File: - should raise an error
2. symlink is provided as an input Directory: - should raise an error
3. Directory with some symlinks is provided as an input Directory: should be ok

**Comment by satra** [comment]
> symlink is provided as an input File: - should raise an error

only if the link is broken, right? there should be no problem with symlinks that are not broken.

**Comment by nicolocin** [comment]
@djarecka - I'm a bit confused cause I thought @satra said we should just skip broken symlinks without raising an error?  I assume that means we allow valid symlinks?  

3 is covered in `test_hash_broken_symlink`.

**Comment by satra** [comment]
in a directory we can skip broken symlinks for hashing

but as a File type, broken symlinks should raise an error

**Comment by djarecka** [comment]
@satra - yes, I meant when the symlink is broken

File: pydra/engine/helpers_file.py [hunk 1]
```diff
@@ -70,22 +70,20 @@ def hash_file(afile, chunk_len=8192, crypto=sha256, raise_notfound=True):
...
+            while True:
+                data = fp.read(chunk_len)
+                if not data:
+                    break
+                crypto_obj.update(data)
+        return crypto_obj.hexdigest()
+    except:
+        if not Path(afile).is_symlink() and raise_notfound:
+            raise FileNotFoundError(f"File {afile} not found.")
         return None
```
Comment (effigies) [inline pydra/engine/helpers_file.py:85]: I would suggest the following:

```suggestion
    try:
        fp = open(afile, "rb")
    except FileNotFoundError:
        if not Path(afile).is_symlink() and raise_notfound:
            raise
        return None    
    crypto_obj = crypto()
    with fp as fp:
        while True:
            data = fp.read(chunk_len)
            if not data:
                break
            crypto_obj.update(data)
    return crypto_obj.hexdigest()
    
```

This keeps the potentially-failing block inside `try:` to a minimum. Using `with` on the resulting object will still have the nice cleanup properties. I also just re-raise the original error, which presumably should have the same information.

Comment (satra) [reply pydra/engine/helpers_file.py:85]: @nicolocin - will you be addressing @effigies review?

**Comment by djarecka** [comment]
@nicolocin - please let me know if you need help with creating the tests I mentioned

**Comment by nicolocin** [comment]
How about when directroy with some _all_ broken symlinks is provided as an input Directory - @djarecka said we shouldn't raise an error, just want to confirm

File: pydra/engine/helpers_file.py (hunk 1 above)
Comment (nicolocin) [reply pydra/engine/helpers_file.py:85]: @effigies @satra - so sorry thought I replied!  Yes this looks better!  Thanks a lot

**Comment by djarecka** [comment]
unfortunately now everything is mixed now in this PR, I'm not sure how to review it now..:(

**Comment by satra** [comment]
@djarecka - should  be fixed now.

File: pydra/engine/helpers_file.py [hunk 2]
```diff
@@ -71,21 +71,21 @@ def hash_file(afile, chunk_len=8192, crypto=sha256, raise_notfound=True):
...
-        return None
-    if not Path(afile).is_file():
-        if raise_notfound:
-            raise RuntimeError('File "%s" not found.' % afile)
-        return None
-
+    try:
+        fp = open(afile, "rb")
+    except FileNotFoundError:
+        if not Path(afile).is_symlink() and raise_notfound:
```
Comment (djarecka) [inline pydra/engine/helpers_file.py:77]: and what if `afile` is not a symlink? it also should raise an error

File: pydra/engine/helpers_file.py [hunk 3]
```diff
@@ -122,10 +122,8 @@ def hash_dir(
     """
     from .specs import LazyField
 
-    if dirpath is None or isinstance(dirpath, LazyField) or isinstance(dirpath, list):
-        return None
     if not Path(dirpath).is_dir():
-        if raise_notfound:
+        if not Path(dirpath).is_symlink() and raise_notfound:  # ignore broken symlinks
```
Comment (djarecka) [inline pydra/engine/helpers_file.py:126]: the same here, what if dir just doesn't exists?

**Comment by djarecka** [comment]
actually, we should have more tests, I'm copying the previous list and adding two more tests:
1. broken symlink is provided as an input File: - should raise an error
1a. broken path is provided as an input File: - should raise an error
2. broken symlink is provided as an input Directory: - should raise an error
2a. broken path is provided as an input Directory: - should raise an error
3. Directory with some broken symlinks is provided as an input Directory: should be ok

File: pydra/engine/helpers_file.py (hunk 2 above)
Comment (nicolocin) [reply pydra/engine/helpers_file.py]: Fixed both hash_file and hash_dir in new commit!

Comment (nicolocin) [reply pydra/engine/helpers_file.py]: Sorry for taking so long....

File: pydra/engine/tests/test_helpers_file.py [hunk 4]
```diff
@@ -184,6 +186,79 @@ def test_get_related_files_noninclusive(_temp_analyze_files):
...
+    Test how broken file paths are handled during hashing
+
+    When `raise_notfound` is True, raise error for broken file paths or
+    broken symlinks. Otherwise return None
+    """
+    # broken file path
+    assert hash_file("broken_file.txt", raise_notfound=False) is None
+    with pytest.raises(FileNotFoundError) as e1:
+        hash_file("broken_file.txt", raise_notfound=True)
+    assert "File doesn't exist" in str(e1)
```
Comment (djarecka) [inline pydra/engine/tests/test_helpers_file.py:200]: not completely sure if this is the only problem with failing tests on travis, but you should probably use `str(e1.value)` in all places you are checking the error message

**Comment by djarecka** [comment]
@nicolocin - thanks for updating. I left the comment - it might fix the problem with travis.

I like your test, but I really think we should also add the tests that I suggested, so it checks if the input is treated properly. Let me know if you're planning to add them or you need help with it.

File: pydra/engine/tests/test_helpers_file.py (hunk 4 above)
Comment (nicolocin) [reply pydra/engine/tests/test_helpers_file.py:200]: Thanks for the suggestion!

**Comment by nicolocin** [comment]
@djarecka  Thanks again for reviewing!   Though I do think I covered all 5 cases in the list in the tests already?  Can you let me know which test case I'm missing?

**Comment by djarecka** [comment]
@nicolocin - your tests cover `hash_file` and `hash_dir` functions alone and these are great unit tests! But I was also thinking about adding a couple of more "functional tests" that tests how these functions work with task's input, and if the proper errors are raised when you provide e.g.,  broken link as an input. 

I've created a [PR](https://github.com/nicolocin/pydra/pull/3) to your fork with an example. I used a broken link as an input to the task. Unfortunately, when I try to calculate a checksum, no error is raised. 
When I run the task the error is raised, but somewhere else (without providing info about the broken link).

**Comment by djarecka** [comment]
@nicolocin - do you think you can come back to it this week and finish?

**Comment by nicolocin** [comment]
Status: working but not if File or Directory is nested in the inputs.  Still WIP

Theses tests are failing
```python
=========================== short test summary info ============================
FAILED pydra/engine/tests/test_boutiques.py::test_boutiques_wf_1[cf-test_brain]
FAILED pydra/engine/tests/test_boutiques.py::test_boutiques_wf_1[cf-test_brain.nii.gz]
FAILED pydra/engine/tests/test_boutiques.py::test_boutiques_wf_1[cf-test_brain.nii]
FAILED pydra/engine/tests/test_boutiques.py::test_boutiques_wf_2[cf-test_brain.nii.gz]
FAILED pydra/engine/tests/test_boutiques.py::test_boutiques_wf_2[cf-test_brain]
FAILED pydra/engine/tests/test_boutiques.py::test_boutiques_wf_2[cf-test_brain.nii]
FAILED pydra/engine/tests/test_dockertask.py::test_wf_docker_3[cf] - FileNotF...
FAILED pydra/engine/tests/test_dockertask.py::test_docker_wf_inputspec_1[cf]
FAILED pydra/engine/tests/test_dockertask.py::test_docker_wf_ndst_inputspec_1[cf]
FAILED pydra/engine/tests/test_dockertask.py::test_docker_wf_state_inputspec_1[cf]
FAILED pydra/engine/tests/test_shelltask.py::test_shell_cmd_inputspec_10_err
FAILED pydra/engine/tests/test_shelltask.py::test_wf_shell_cmd_3[cf] - FileNo...
FAILED pydra/engine/tests/test_specs.py::test_input_file_hash_3 - FileNotFoun...
FAILED pydra/engine/tests/test_specs.py::test_input_file_hash_4 - FileNotFoun...
FAILED pydra/engine/tests/test_tasks_files.py::test_text_task - _pickle.Pickl...
FAILED pydra/engine/tests/test_tasks_files.py::test_broken_task - AttributeEr...
FAILED pydra/engine/tests/test_tasks_files.py::test_broken_file_link - _pickl...
FAILED pydra/engine/tests/test_tasks_files.py::test_broken_dir_link2 - Failed...
FAILED pydra/engine/tests/test_workflow.py::test_wf_resultfile_1[cf] - FileNo...
FAILED pydra/engine/tests/test_workflow.py::test_wf_resultfile_2[cf] - FileNo...
FAILED pydra/engine/tests/test_workflow.py::test_wf_resultfile_3[cf] - FileNo...
FAILED pydra/utils/tests/test_messenger.py::test_file_messenger - assert 1 == 2
= 22 failed, 746 passed, 31 skipped, 5 xfailed, 8 warnings, 18 rerun in 579.13s (0:09:39) =
Error: Process completed with exit code 1.

```

**Comment by nicolocin** [comment]
Starting over on a new branch #372

**Pull request closed without merging**